File: backend/models/offer.py
from typing import Any, Literal

from pydantic import BaseModel, Field

from backend.models.common import Money
from backend.models.trust import TrustScore


class Offer(BaseModel):
    offer_id: str
    source: Literal["ebay", "serpapi"]
    title: str
    description: str | None = None
    price: Money
    url: str
    image_urls: list[str] = []
    seller_id: str
    seller_name: str
    free_shipping: bool = False
    estimated_delivery_days: int | None = None
    condition: Literal["new", "used", "refurbished", "unknown"] = "unknown"
    raw_attributes: dict[str, Any] = {}


class ScoredOffer(Offer):
    trust_score: TrustScore


class RankedOffer(ScoredOffer):
    composite_score: float = Field(ge=0.0, le=100.0)
    rank: int = Field(ge=1)
    price_score: float = Field(ge=0.0, le=25.0)
    relevance_score: float = Field(ge=0.0, le=20.0)
    rating_score: float = Field(ge=0.0, le=15.0)
    shipping_score: float = Field(ge=0.0, le=5.0)

File: frontend/src/types/offer.ts
export type TrustLevel = 'LOW_RISK' | 'MEDIUM_RISK' | 'HIGH_RISK' | 'INSUFFICIENT_DATA'

export interface TrustSignals {
  rating: number | null
  review_count: number | null
  account_age_days: number | null
  has_return_policy: boolean | null
  fulfilled_orders: number | null
  feedback_percentage: number | null
}

export interface TrustScore {
  score: number
  level: TrustLevel
  signals: TrustSignals
  explanation: string
  data_source: 'ebay_api' | 'insufficient'
}

export interface Money {
  amount: number
  currency: string
}

export interface Offer {
  offer_id: string
  source: 'ebay' | 'serpapi'
  title: string
  description: string | null
  price: Money
  url: string
  image_urls: string[]
  seller_id: string
  seller_name: string
  free_shipping: boolean
  estimated_delivery_days: number | null
  condition: 'new' | 'used' | 'refurbished' | 'unknown'
}

export interface ScoredOffer extends Offer {
  trust_score: TrustScore
}

export interface RankedOffer extends ScoredOffer {
  composite_score: number
  rank: number
  price_score: number
  relevance_score: number
  rating_score: number
  shipping_score: number
}
